app/uk/gov/hmrc/gform/graph/Recalculation.scala

```diff
@@ -331,6 +357,25 @@ class Recalculation[F[_]: Monad, E](
         .flatMap(_.includeIf)
     }
 
+  private def isHiddenByValidIf(
+    fcId: FormComponentId,
+    evaluationResults: EvaluationResults,
+    recData: RecData[SourceOrigin.OutOfDate],
+    retrievals: MaterialisedRetrievals,
+    booleanExprResolver: BooleanExprResolver,
+    evaluationContext: EvaluationContext
+  )(implicit formModel: FormModel[Interim]): StateT[F, RecalculationState, Boolean] = {
+    val validIfs: List[ValidIf] = formModel.allValidIfs.filter(_._2.id === fcId).flatMap(_._1)
+
+    validIfs
+      .traverse(validIf =>
+        evaluateValidIf(evaluationResults, recData, retrievals, booleanExprResolver, evaluationContext) {
+          Some(validIf)
+        }
+      )
+      .map(_.foldLeft(false)(_ && _))
```

Comment:
Name of this method `isHiddenByValidIf` doesn't sound right. `validIf` cannot hide anything.

Also this line (376) will always evaluate to `false`, unless I'm missing something.

I think (and I may be wrong) all we want is 
```
formModel.allValidIfs.traverse(validIf =>
  evaluateValidIf(evaluationResults, recData, retrievals, booleanExprResolver, evaluationContext) {
    Some(validIf)
  }
)
```

And combine it with result of 
```
graphLayer.foldMapM {
  ...
}
 ```
 
 
Also `evaluateValidIf` and `evaluateIncludeIf` are almost the same, we should refactor common code.